Improve ExistingFsxNetworkingValidator Validator

hanwen-pcluste opened the pull request:
1. Only display one error message if multiple ports of a single file system are absent
2. Improve the validator to check UDP protocol when necessary. This was not implementated because we had no time
3. Refactor the code avoid function too complex code linters error

Signed-off-by: Hanwen <hanwenli@amazon.com>

### Tests
Tests in config/pcluster3.yaml have been passed
Manually tested the validator works as expected when the ports are not completed

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hanwen-cluster committed:
Improve ExistingFsxNetworkingValidator Validator

1. Only display one error message if multiple ports of a single file system are absent
2. Improve the validator to check UDP protocol when necessary. This was not implementated because we had no time
3. Refactor the code avoid function too complex code linters error

Signed-off-by: Hanwen <hanwenli@amazon.com>

gmarciani commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -438,9 +440,16 @@ def _check_sg_rules_for_port(rule, port_to_check):
-    if (ip_protocol in ["tcp", "6"]) and (from_port <= port_to_check <= to_port):
+    # Add protocol number in addition to the protocol name
+    if protocol == "tcp":
+        expected_protocol = [protocol, "6"]
```
Why do we need these magic numbers: 6 and 17? Who is in control of these numbers? Could these numbers be dynamically retrieved to avoid tech debts?

hanwen-pcluste replied:
According to the [boto3 doc](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_security_groups), Either the string or the number is included in the boto3 result.

To make it dynamically, we need to include a Python dependencies https://pypi.org/project/sockets/. I believe we don't want to include the dependencies just for these numbers?

gmarciani replied:
For sure we don't want a new external deps. LEt's go with the proposed solution. thanks

gmarciani commented on cli/src/pcluster/validators/cluster_validators.py:
```diff
@@ -505,23 +514,40 @@ def _check_file_systems(self, are_all_security_groups_customized, file_systems,
+                if _check_in_out_access(
+                    sg_ids,
+                    port=port,
+                    is_cidr_optional=are_all_security_groups_customized,
```
Why `is_cidr_optional=are_all_security_groups_customized`? I don't get why the CIDR is optional when all SGs are customized.

hanwen-pcluste replied:
If all security groups are customized, CIDR specification is not required. Because custom security groups can allow traffics by specifying security group id of themselves or others.
On the contrary, if cluster nodes use ParallelCluster managed security group, the security groups of the file system has to have CIDR specification to enable traffics from the file system to the nodes

hanwen-cluster committed:
Clean up _register_storage_validators method

1. Make `existing_fsx` a set instead of a list to avoid duplication. Because it is possible to specify different volumes from the same FSx Ontap/OpenZFS. But we don't need `existing_fsx` to contain duplicate items

2. Change validation logic from multiple `if` to multiple `elif` to make the code more readable

Signed-off-by: Hanwen <hanwenli@amazon.com>

gmarciani commented on cli/src/pcluster/config/cluster_config.py:
```diff
@@ -1254,7 +1255,7 @@ def _register_storage_validators(self):
                         new_storage_count["efs"] += 1
             self._register_validator(
                 ExistingFsxNetworkingValidator,
```
Minor: Please keep track in our backlog that we must cover these kind of changes in unit tests like `test_validators_are_called_with_correct_argument` to be sure that expected validators are called with the correct arguments

hanwen-pcluste committed:
Merge branch 'develop' into jul11improveerrormessage
